smart-py/objkt_swap_v3_test.py

```diff
@@ -531,3 +541,442 @@ def test_pause_contract():
+        valid=False
+    )
+
+    # No swap was added
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == False)
+    scenario.verify(marketplaceV3.data.counter == 0)
+
+    # trying to swap someone elses objkt as admin must fail
+    scenario += marketplaceV3.swap(
+        fa2=objkt.address,
+        objkt_id=152,
+        objkt_amount=1,
+        xtz_per_objkt=sp.mutez(10000),
+        royalties=100,
+        creator=artist1.address
+    ).run(
+        sender=admin,
+        valid=False
+    )
+
+    # No swap was added
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == False)
+    scenario.verify(marketplaceV3.data.counter == 0)
+
+    # Cannot swap 0 items
+    scenario += marketplaceV3.swap(
+        fa2=objkt.address,
+        objkt_id=152,
+        objkt_amount=0,
+        xtz_per_objkt=sp.mutez(10000),
+        royalties=100,
+        creator=artist1.address
+    ).run(
+        sender=artist1,
+        valid=False
+    )
+
+    # swap was not added
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == False)
+    scenario.verify(marketplaceV3.data.counter == 0)
+
+    # Successfully swap
+    scenario += marketplaceV3.swap(
+        fa2=objkt.address,
+        objkt_id=152,
+        objkt_amount=1,
+        xtz_per_objkt=sp.mutez(10000),
+        royalties=100,
+        creator=artist1.address
+    ).run(
+        sender=artist1,
+        valid=True
+    )
+
+    # swap was added
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == True)
+    scenario.verify(marketplaceV3.data.swaps.contains(1) == False)
+    scenario.verify(marketplaceV3.data.counter == 1)
+
+    # second swap should now fail because there is only 1 edition available
+    scenario += marketplaceV3.swap(
+        fa2=objkt.address,
+        objkt_id=152,
+        objkt_amount=1,
+        xtz_per_objkt=sp.mutez(10000),
+        royalties=100,
+        creator=artist1.address
+    ).run(
+        sender=artist1,
+        valid=False
+    )
+
+    # no swap was added
+    scenario.verify(marketplaceV3.data.swaps.contains(0) == True)
+    scenario.verify(marketplaceV3.data.swaps.contains(1) == False)
+    scenario.verify(marketplaceV3.data.counter == 1)
+
+    # Mint a multi edition from a second OBJKT
+    scenario += marketplaceV1.mint_OBJKT(
+        address=artist2.address,
+        amount=10,
+        metadata=sp.pack("ipfs://fff"),
+        royalties=100
+    ).run(
+        sender=artist2
+    )
+
+    # objkt was added
+    scenario.verify(marketplaceV1.data.objkt_id == 154)
+
+    # Artist 2 updates the new objkt to be managed by marketplace v3
+    scenario += objkt.update_operators(
+        [sp.variant("add_operator", objkt.operator_param.make(
+            owner=artist2.address,
+            operator=marketplaceV3.address,
+            token_id=153))]).run(sender=artist2)
+
+    # Fail to swap second objkt as second artist when royalty mismatch
+    #
+    # TODO this should fail because there is a royalty mismatch
```

Comment:
@jagracar is it possible for swap contract to check royalties on the objkt before allowing the swap?

Reply:
Nope, as discussed per DM, it's not possible to get the royalties and creator info from the v1 contract. This can only be done on the UI using the indexer or tzkt calls to get that information before submitting the swap. 
This is what is done current. People swapping in H=N cannot enter the royalties and creator info, the UI is doing it for them.
Of course, people could BCD to create the swap and put some different royalties and creator. That cannot be avoided right now. But the site again has a solution for that. Swaps that don't use the original royalties and creators are not displayed on the site, so people cannot collect those swaps using the site. They are invisible.